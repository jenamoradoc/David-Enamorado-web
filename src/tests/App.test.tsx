import {
  cleanup,
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import App from "../App";

describe("Testing <App />", () => {
  // afterEach(cleanup);

  test("Should render the initial content", () => {
    const { container } = render(<App />);

    expect(container).toMatchSnapshot();
    expect(container).toBeInTheDocument();
  });

  test("should show aside container", () => {
    render(<App />);

    expect(screen.getByRole("complementary")).toBeInTheDocument();
  });

  test("should show the input field", () => {
    render(<App />);

    expect(screen.getByRole("textbox")).toBeInTheDocument();
  });

  test("should be empty the input field", () => {
    render(<App />);

    expect((screen.getByRole("textbox") as HTMLInputElement).value).toBe("");
    expect((screen.getByRole("textbox") as HTMLInputElement).value.length).toBe(
      0
    );
  });

  test("should show a Submit button", () => {
    render(<App />);
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  test("should change the input value", async () => {
    render(<App />);

    userEvent.type(screen.getByRole("textbox"), "pedro");

    expect((screen.getByRole("textbox") as HTMLInputElement).value).toBe(
      "pedro"
    );
  });

  test("should show results when the input field is changed and the button is clicked", async () => {
    const { container } = render(<App />);

    userEvent.type(screen.getByRole("textbox"), "pedro");
    userEvent.click(screen.getByText(/submit/i));

    expect(screen.getByTestId("spinner")).toBeInTheDocument();

    await waitForElementToBeRemoved(() => screen.queryByTestId("spinner"));

    await screen.findByRole("list");
    expect(screen.getAllByRole("listitem").length).toBeGreaterThan(0);

    expect(container).toMatchSnapshot();
  });

  test("should display the previous and next page buttons", async () => {
    render(<App />);

    userEvent.type(screen.getByRole("textbox"), "pedro");
    userEvent.click(screen.getByText(/submit/i));

    expect(screen.getByTestId("spinner")).toBeInTheDocument();

    await waitFor(() => screen.findByRole("list"));
    expect(screen.getAllByRole("listitem").length).toBeGreaterThan(0);

    expect(screen.getByText("«")).toBeTruthy();
    expect(screen.getByText("»")).toBeTruthy();
  });

  test("should show the next results when the Next Page button is clicked", async () => {
    const { container } = render(<App />);

    userEvent.type(screen.getByRole("textbox"), "pedro");
    userEvent.click(screen.getByText(/submit/i));

    expect(screen.getByTestId("spinner")).toBeInTheDocument();

    await waitForElementToBeRemoved(() => screen.queryByTestId("spinner"));

    if (
      screen.getAllByRole("listitem").length === 0 &&
      screen.getByText(/something went wrong/i)
    ) {
      return;
    }

    await screen.findByRole("list");
    expect(screen.getAllByRole("listitem").length).toBeGreaterThan(0);

    userEvent.click(screen.getByText("»"));
    expect(screen.getByTestId("spinner")).toBeInTheDocument();
    await waitForElementToBeRemoved(screen.queryByTestId("spinner"));

    await waitFor(() => screen.findByRole("list"));
    expect(screen.getAllByRole("listitem").length).toBeGreaterThan(0);

    expect(container).toMatchSnapshot();
  });
});
