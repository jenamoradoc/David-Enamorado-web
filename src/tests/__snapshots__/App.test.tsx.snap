// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <App /> Should render the initial content 1`] = `
<div>
  <div
    class="container"
  >
    <aside
      class="search__form-container"
    >
      <form
        class="search__form"
      >
        <input
          class="search__form-input"
          placeholder="Enter a username"
          type="text"
          value=""
        />
        <button
          class="search__form-btn"
          type="submit"
        >
          Submit
        </button>
      </form>
    </aside>
    <div />
  </div>
</div>
`;

exports[`Testing <App /> should show results when the input field is changed and the button is clicked 1`] = `
<div>
  <div
    class="container"
  >
    <aside
      class="search__form-container"
    >
      <form
        class="search__form"
      >
        <input
          class="search__form-input"
          placeholder="Enter a username"
          type="text"
          value="pedro"
        />
        <button
          class="search__form-btn"
          type="submit"
        >
          Submit
        </button>
      </form>
    </aside>
    <div>
      <select
        class="results__sort"
      >
        <option
          disabled=""
          value=""
        >
          --Choose an option
        </option>
        <option
          value="best match"
        >
          Best Match
        </option>
        <option
          value="followers"
        >
          Followers
        </option>
        <option
          value="repositories"
        >
          Repositories
        </option>
        <option
          value="joined"
        >
          Joined
        </option>
      </select>
      <ul
        class="results__list"
      >
        <li
          class="results__list-item"
        >
          <img
            alt="pedro"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/2248?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedro
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedrovgs"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/4030704?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedrovgs
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedronauck"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/2029172?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedronauck
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedromassango"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/33294549?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedromassango
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="Pedromdsn"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/57453260?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            Pedromdsn
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedrouid"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/10136079?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedrouid
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedroslopez"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/4368928?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedroslopez
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedromarins"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/215004?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedromarins
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedrofracassi"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/25179120?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedrofracassi
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
      </ul>
      <div
        class="buttons__container"
      >
        <button
          disabled=""
          style="border-radius: 5px 0 0 5px;"
        >
          «
        </button>
        <button
          data-testid="1"
          disabled=""
        >
          1
        </button>
        <button
          data-testid="2"
        >
          2
        </button>
        <button
          data-testid="3"
        >
          3
        </button>
        <button
          data-testid="4"
        >
          4
        </button>
        <button
          data-testid="5"
        >
          5
        </button>
        <button
          data-testid="6"
        >
          6
        </button>
        <button
          data-testid="7"
        >
          7
        </button>
        <button
          data-testid="8"
        >
          8
        </button>
        <button
          data-testid="9"
        >
          9
        </button>
        <button
          style="border-radius: 0 5px 5px 0;"
        >
          »
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing <App /> should show the next results when the Next Page button is clicked 1`] = `
<div>
  <div
    class="container"
  >
    <aside
      class="search__form-container"
    >
      <form
        class="search__form"
      >
        <input
          class="search__form-input"
          placeholder="Enter a username"
          type="text"
          value="pedro"
        />
        <button
          class="search__form-btn"
          type="submit"
        >
          Submit
        </button>
      </form>
    </aside>
    <div>
      <select
        class="results__sort"
      >
        <option
          disabled=""
          value=""
        >
          --Choose an option
        </option>
        <option
          value="best match"
        >
          Best Match
        </option>
        <option
          value="followers"
        >
          Followers
        </option>
        <option
          value="repositories"
        >
          Repositories
        </option>
        <option
          value="joined"
        >
          Joined
        </option>
      </select>
      <ul
        class="results__list"
      >
        <li
          class="results__list-item"
        >
          <img
            alt="pedrualves"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/1184178?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedrualves
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedrobaeza"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/7165771?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedrobaeza
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedrofranceschi"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/114711?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedrofranceschi
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pepicrft"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/663605?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pepicrft
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pmarkun"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/682827?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pmarkun
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedroabreudev"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/85113076?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedroabreudev
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedrobergamini"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/41773103?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedrobergamini
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedrommcarrasco"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/26525777?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedrommcarrasco
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
        <li
          class="results__list-item"
        >
          <img
            alt="pedrohcdo"
            class="results__list-item-image"
            draggable="false"
            src="https://avatars.githubusercontent.com/u/5511316?v=4"
          />
          <strong
            class="results__list-item-login"
          >
            pedrohcdo
          </strong>
          <span
            class="results__list-item-type"
          >
            User
          </span>
        </li>
      </ul>
      <div
        class="buttons__container"
      >
        <button
          style="border-radius: 5px 0 0 5px;"
        >
          «
        </button>
        <button
          data-testid="1"
        >
          1
        </button>
        <button
          data-testid="2"
          disabled=""
        >
          2
        </button>
        <button
          data-testid="3"
        >
          3
        </button>
        <button
          data-testid="4"
        >
          4
        </button>
        <button
          data-testid="5"
        >
          5
        </button>
        <button
          data-testid="6"
        >
          6
        </button>
        <button
          data-testid="7"
        >
          7
        </button>
        <button
          data-testid="8"
        >
          8
        </button>
        <button
          data-testid="9"
        >
          9
        </button>
        <button
          style="border-radius: 0 5px 5px 0;"
        >
          »
        </button>
      </div>
    </div>
  </div>
</div>
`;
